exporter/splunkhecexporter/logdata_to_splunk.go

```diff
@@ -15,29 +15,156 @@
 package splunkhecexporter
 
 import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
 	"time"
 
+	"go.opentelemetry.io/collector/consumer/consumererror"
 	"go.opentelemetry.io/collector/consumer/pdata"
 	"go.opentelemetry.io/collector/translator/conventions"
 	"go.uber.org/zap"
 
 	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/splunk"
 )
 
-func logDataToSplunk(logger *zap.Logger, ld pdata.Logs, config *Config) []*splunk.Event {
-	var splunkEvents []*splunk.Event
-	rls := ld.ResourceLogs()
-	for i := 0; i < rls.Len(); i++ {
-		ills := rls.At(i).InstrumentationLibraryLogs()
-		for j := 0; j < ills.Len(); j++ {
-			logs := ills.At(j).Logs()
-			for k := 0; k < logs.Len(); k++ {
-				splunkEvents = append(splunkEvents, mapLogRecordToSplunkEvent(logs.At(k), config, logger))
+// eventsBuf is a buffer of JSON encoded Splunk events.
+// The events are created from LogRecord(s) where one event maps to one LogRecord.
+type eventsBuf struct {
+	buf *bytes.Buffer
+	// index is the eventIndex of the 1st event in buf.
+	index *eventIndex
+	err   error
+}
+
+// The index of an event composed of indices of the event's LogRecord.
+type eventIndex struct {
+	// Index of the LogRecord slice element from which the event is created.
+	log int
+	// Index of the InstrumentationLibraryLogs slice element parent of the LogRecord.
+	lib int
+	// Index of the ResourceLogs slice element parent of the InstrumentationLibraryLogs.
+	src int
+}
+
+type logDataWrapper struct {
+	*pdata.Logs
+}
+
+func (ld *logDataWrapper) eventsInChunks(logger *zap.Logger, config *Config) (chan *eventsBuf, context.CancelFunc) {
+	ctx, cancel := context.WithCancel(context.Background())
+	eventsCh := make(chan *eventsBuf)
+
+	go func() {
+		defer close(eventsCh)
+
+		// event buffers a single event.
+		event := new(bytes.Buffer)
+		encoder := json.NewEncoder(event)
+
+		// events buffers events up to the max content length.
+		events := &eventsBuf{buf: new(bytes.Buffer)}
+
+		rl := ld.ResourceLogs()
+		for i := 0; i < rl.Len(); i++ {
+			ill := rl.At(i).InstrumentationLibraryLogs()
+			for j := 0; j < ill.Len(); j++ {
+				l := ill.At(j).Logs()
+				for k := 0; k < l.Len(); k++ {
+					select {
+					case <-ctx.Done():
+						return
+					default:
+						if err := encoder.Encode(mapLogRecordToSplunkEvent(l.At(k), config, logger)); err != nil {
+							eventsCh <- &eventsBuf{buf: nil, index: nil, err: consumererror.Permanent(err)}
+							return
+						}
+						event.WriteString("\r\n\r\n")
+
+						// The size of an event must be less than or equal to max content length.
+						if config.MaxContentLength > 0 && event.Len() > config.MaxContentLength {
+							err := fmt.Errorf("found a log event bigger than max content length (event: %d bytes, max: %d bytes)", config.MaxContentLength, event.Len())
+							eventsCh <- &eventsBuf{buf: nil, index: nil, err: consumererror.Permanent(err)}
```

Comment:
That's unfortunate. Is there a way to try and zip the event to see if it fits under max content length?